admin/services/access-arrangements.service.js

```diff
@@ -10,4 +16,65 @@ accessArrangementsService.getAccessArrangements = async () => {
   return accessArrangementsDataService.sqlFindAccessArrangements()
 }
 
-module.exports = accessArrangementsService
+/**
+ * Process access arrangements for single pupil
+ * @param {Object} requestData
+ * @param {Number} dfeNumber
+ * @param {Number} userId
+ * @returns {Object}
+ */
+accessArrangementsService.submit = async (requestData, dfeNumber, userId) => {
+  const validationError = accessArrangementsValidator.validate(requestData)
+  if (validationError.hasError()) {
+    throw validationError
+  }
+  const { pupilUrlSlug, accessArrangements: accessArrangementsCodes, questionReaderReason: questionReaderReasonCode } = requestData
+  let pupilAccessArrangements = R.clone(requestData)
+  const accessArrangements = await accessArrangementsDataService.sqlFindAccessArrangementsByCodes(accessArrangementsCodes)
+  let questionReaderReason
+  if (accessArrangements.length === 0) {
+    throw new Error('No access arrangements found')
+  }
+  pupilAccessArrangements['accessArrangements_ids'] = JSON.stringify(accessArrangements.map(aa => aa.id))
+  const pupil = await pupilDataService.sqlFindOneBySlugAndSchool(pupilUrlSlug, dfeNumber)
+  if (!pupil) {
+    throw new Error('Pupil url slug does not match a pupil record')
+  }
+  pupilAccessArrangements['pupil_id'] = pupil.id
+  pupilAccessArrangements['recordedBy_user_id'] = userId
+  if (!pupilAccessArrangements.accessArrangements.includes('ITA')) {
+    pupilAccessArrangements = R.omit(['inputAssistanceInformation'], pupilAccessArrangements)
+  }
+  if (pupilAccessArrangements.questionReaderReason !== 'OTH') {
+    pupilAccessArrangements = R.omit(['questionReaderOtherInformation'], pupilAccessArrangements)
+  }
+  if (pupilAccessArrangements.accessArrangements.includes('QNR')) {
+    questionReaderReason = await questionReaderReasonsDataService.sqlFindQuestionReaderReasonByCode(questionReaderReasonCode)
+  }
+  if (questionReaderReason && questionReaderReason.id) {
+    pupilAccessArrangements['questionReaderReasons_id'] = questionReaderReason.id
+  }
+  pupilAccessArrangements = R.omit(['_csrf', 'accessArrangements', 'questionReaderReason', 'pupilUrlSlug'], pupilAccessArrangements)
+  return accessArrangementsService.save(pupilAccessArrangements, pupil)
+}
+
+/**
+ * Save access arrangements for single pupil
+ * @param {Object} pupilAccessArrangements
+ * @param {Object} pupil
+ * @returns {Object}
+ */
+
+accessArrangementsService.save = async (pupilAccessArrangements, pupil) => {
+  const pupilAccessArrangement = await pupilAccessArrangementsDataService.sqlFindPupilAccessArrangementsByPupilId(pupil.id)
+  if (pupilAccessArrangement && pupilAccessArrangement['pupil_id']) {
+    // update
+    await pupilAccessArrangementsDataService.sqlUpdate(pupilAccessArrangements)
+  } else {
+    // create
+    await pupilAccessArrangementsDataService.sqlCreate(pupilAccessArrangements)
+  }
+  return pupil
```

Comment:
Is this return actually useful?

Reply:
@jon-shipley I need to use on controller level the id, foreName and lastName of the pupil for the highlight message.
Perhaps I could return a simplified object with just these three properties.
Any objections?